### ISSUE TITLE
Add ramp up/down for matmul

### ISSUE THREAD

**Issue by TT-BrianLiu** [tenstorrent/tt-metal#11308]
# Feature
For di/dt testing, we want to be able to configure matmul to ramp up/down at the beginning and end to smooth out current draw. This is an example of what it will look look:
- loop 1: 8 cores active
- loop 2: 16 cores active
- loop 3: 32 cores active
- loop 4 to N-3: all cores active
- loop N-2: 32 cores active
- loop N-1: 16 cores active
- loop N: 8 cores active

For implementation, we will group matmul cores together and cores within a group sync with another (ie. only one core within the group can run compute). As we increase active cores, we will "shrink" the group size and vice versa. Ideally, we can configure two things:
- Maximum number of active cores at the beginning and end of the matmul
- How fast we ramp up/down (ie. by factor of 2, 4, etc...)

# Implementation
Branch: `bliu/didt`

For sharded 1D or 2D matmuls, there are two orthogonal features that you can try out by adding these settings in your environment variables:
```
TT_MATMUL_RAMP_CONFIG='{"enable_dummy_loops": true, "enable_ramp": false, "initial_cores": 8, "ramp_multiple": 2}'
```
- `enable_dummy_loops`: If true, insert additional loops of dummy `matmul_block` compute at the beginning and end of actual matmul loops to gradually turn on/off cores. In the middle, run actual matmul at full capacity.
- `enable_ramp`: If true, puts the actual matmul loops into a start, middle, and end phase where cores are grouped and are gradually turned on/off within the group.
- `initial_cores`: How many cores to start/end with. This is used to determine the group size. ie. If 64 cores total, initial cores of 8 means group size of 8; if 56 cores total, initial cores of 8 means group size of 7.
- `ramp_multiple`: How fast we turn on/off cores. If initial cores of 8 and ramp multiple of 2, you the second loop 16 cores will be running, etc... This applies to both `enable_dummy_loops` and `enable_ramp` feature, although they are implemented slightly differently.

NOTE: You can probably run both features together, but I haven't tested that.

# Test Results with PMON Capture 
Test are run on `sjc-snva-t3005` with chip 0, since this is the chip that consistently hangs in the 8-chip hang repro.

PMON visualization:
<img width="1043" alt="image" src="https://github.com/user-attachments/assets/f04920d8-f191-46ff-8f2b-997d41cc31c5">

PMON fit:
<img width="508" alt="image" src="https://github.com/user-attachments/assets/9921e837-7665-4305-9102-3a38062439c1">

## Enable ramp tests
I modified the test to run only 5000 loops for captures instead of 10000. Test command is this (`initial_cores` is varied for experiments):
```
TT_MATMUL_RAMP_CONFIG='{"enable_dummy_loops": false, "enable_ramp": true, "initial_cores": 8, "ramp_multiple": 2}' 
WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest tests/didt/test_sharded_ff1.py::test_specific_chip_reproduce_matmul_2d_hang_t3000[logical_chip0]
```

Baseline (ie. don't set anything related to `TT_MATMUL_RAMP_CONFIG`): 
<img width="1102" alt="image" src="https://github.com/user-attachments/assets/67228522-28e9-4337-b16a-8b2914e72b2d">

Baseline with stagger (`TT_ENABLE_MATMUL_STAGGER=1`):
<img width="1132" alt="image" src="https://github.com/user-attachments/assets/470db78f-d8de-44b8-9f5e-1f1e075a05e4">

Initial cores 1:
<img width="1131" alt="image" src="https://github.com/user-attachments/assets/0ca07ebf-5fb0-4d38-a5e7-833bf7ae56b0">

Initial cores 8:
<img width="1113" alt="image" src="https://github.com/user-attachments/assets/1e74ac23-bb95-4c5c-9b89-50bae4d8ae2c">

Initial cores 32:
<img width="1132" alt="image" src="https://github.com/user-attachments/assets/b55ecb15-14f8-43a7-ac22-58601e9016d0">

## Enable dummy loops tests
IMPORTANT EDIT: I think I may have messed up my implementation of these dummy loops. Each core acts independently of one another. So apart from some basic overhead of arithmetic on risc, the cores may be starting all at roughly the same time, with the "staggered" cores turning off near the end of ramp up. During ramp down, this should be fine as is since cores all start off doing work and slowly drop off. To properly implement the ramp up, we need to do some syncing between cores. We can do this by using simple mcasting:
- Add the same dummy loops to reader kernel of all cores
- Add syncs between reader and compute
- Control the loops by using core 0 (since it should always be running every dummy loop). Core 0 is sender, all other cores are receivers. Every time core 0 is about to start compute, mcast VALID to all other cores to let them know it is the next loop.

### Single chip tests
Same as before except I turn on `enable_dummy_loops` and turn off `enable_ramp`:
```
TT_MATMUL_RAMP_CONFIG='{"enable_dummy_loops": true, "enable_ramp": false, "initial_cores": 8, "ramp_multiple": 2}' 
WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest tests/didt/test_sharded_ff1.py::test_specific_chip_reproduce_matmul_2d_hang_t3000[logical_chip0]
```

Baseline:
<img width="1082" alt="image" src="https://github.com/user-attachments/assets/9dd7f5b1-10bc-4e0a-8eb3-00280c125bf1">

Initial cores 1:
<img width="1103" alt="image" src="https://github.com/user-attachments/assets/6264cdf2-0b32-4a22-928f-7d09c9360d0e">

Initial cores 8:
<img width="1107" alt="image" src="https://github.com/user-attachments/assets/2139e1d0-272c-4979-baad-475798f814f1">

### Multichip tests
Intent here is to see if voltage is different when running multi-chip workloads vs. single chip. PMON by default captures from chip 0, which happens to be the chip I am running my single chip tests on. If you are running single chip tests on another chip, you need to switch which device you are running PMON from.

Same environment setup as single chip tests, but switch the pytest option to the 8 chips one:
```
TT_MATMUL_RAMP_CONFIG='{"enable_dummy_loops": true, "enable_ramp": false, "initial_cores": 8, "ramp_multiple": 2}' 
WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml pytest tests/didt/test_sharded_ff1.py::test_reproduce_matmul_2d_hang[8chips-ff1-hang]
```

Baseline:
<img width="1121" alt="image" src="https://github.com/user-attachments/assets/59ac2bbc-89fb-421e-a7a9-d571913187fa">

Initial cores 1:
<img width="1135" alt="image" src="https://github.com/user-attachments/assets/a5f9c2fb-8bcb-4d1b-b1f2-1a80c3b56adb">

Initial cores 8:
<img width="1120" alt="image" src="https://github.com/user-attachments/assets/d0a93a04-a01e-4bde-a162-89b4ff2d6b17">

**Comment by TT-BrianLiu** [comment]
TLDR:
- `enable_ramp`: This features makes cores alternate compute for initial and tail-end matmul loops, which may not be ideal for power usage. So I started with a simpler approach of adding dummy loops at the beginning and end.
- `enable_dummy_loops`: Feature was not implemented correctly. We need to sync the dummy loops across cores with mcasting:
  - Add the same dummy loops to reader kernel of all cores
  - Add syncs between reader and compute
  - Control the loops by using core 0 (since it should always be running every dummy loop). Core 0 is sender, all other cores are receivers. Every time core 0 is about to start compute, mcast VALID to all other cores to let them know it is the next loop.
- multi-chip tests: Voltage usgae for multi-chip vs. single chip tests looks different. Needs more testing.

**Comment by s-jovic** [comment]
The example used in the issue description (ff1 matmul) triggers AICLK throttling, so nominal voltage is not constant. I repeated the experiment (enable ramp experiment, not the dummy loops one) with fixed voltage and AICLK (850mV and 650MHz). I did the experiment on both N150 and N300 (both chips), and observed similar behavior, so I am adding results from N150 only (T3008 board 0).

The biggest droops are always the first time all 64 cores start working, regardless of whether we had some ramp up work before or not. Since we are doing synchronization between cores in ramp-up iterations, they all more or less finish at the same time, and when all 64 start working we have similar droop as we do in the baseline case. Seems that **the fact that we had some workload running on less cores before we turn on all 64 of them doesn't reduce the droop meaningfully**. @TT-BrianLiu @davorchap @ttmtrajkovic 

To confirm this, in each case I took 6 captures with PMON script which sum up to ~200-300 iterations of matmul, and I didn't see a meaningful reduction in the droop nor any correlation between the ramp-up step/starting point and droop we see when all 64 cores finally start. 

**Baseline case**
The red square shows one iteration of FF1 matmul (48 blocks). The biggest droop is always when all 64 start doing the first block, although it isn't the same for every iteration. Minimum of the iterations I captured is 782.6mV. 

![Screenshot 2024-09-23 at 15 04 32](https://github.com/user-attachments/assets/afa6e393-1eca-4258-87e5-bf6ed4f834c4)

**Ramp-up starting with 32 cores**
The red square shows 2 phases of 32-core iteration. Still, the biggest droop that happens is the one when all 64 cores start, and the minimal one I saw in captured iterations is 783.8mV.  

![Screenshot 2024-09-23 at 15 10 01](https://github.com/user-attachments/assets/776b8924-21d6-485a-a72a-2a2705522525)

Ramp ups starting with 8 or 16 with step 2 behave in a similar way. Ramp-up starting with 16 cores and step 4 showed worse droop - minimal was around 780mV. 

**Dummy loops**
Regarding the dummy loops solution, the results show that we increase the overshoot problem when running these loops in the beginning. This is with current solution. Regarding sync you mentioned, where all cores in group would sync for each new iteration - I am not sure I understand the reasoning. With sync, seems to me that all cores would turn off at the same time again before the 64-core matmul starts. The difference I see is that we would switch from 32+ cores to 64 in the last step, instead of going 32->64, but I am probably missing something - let me know. @TT-BrianLiu 


![Screenshot 2024-09-23 at 15 18 28](https://github.com/user-attachments/assets/095ef66a-e88c-46a0-b8ae-30b068c34a22)
